SAP/spartacus
# fix: CXSPA-7778 RenderingCache is initialized too early in OptimizedSsrEngine

This PR contains a fix for an issue that `RenderingCache` is initialized too early in `OptimizedSsrEngine`, before optimization options provided to the engine's constructor are merged with the default ones. Because of that `RenderingCache` had no notion about default optimization options.

Note: Without a fix for this bug, a solution provided in https://github.com/SAP/spartacus/pull/18467 is not effective.

QA steps:

- go to `ssr-optimization-options.ts` and change default value of `cacheSize` to `1`
- in `server.ts` add `cache: true` to `ssrOptions` to enable caching`
- run `npm run dev:ssr`
- open `http://localhost:4200/electronics-spa/en/USD/` and verify in server's logs whether page has been rendered:
  ```
  {
    "message": "Request is resolved with the SSR rendering result (/electronics-spa/en/USD/)",
    / ... /
  }
  ```
- refresh the page and check in logs whether a page has been taken from the cache:
  ```
  {
    "message": "Render from cache (/electronics-spa/en/USD/)",
    / ... /
  }
  ```
- go to `http://localhost:4200/electronics-spa/en/USD/Brands/all/c/brands` and verify whether a page has been rendered
  ```
  {
    "message": "Request is resolved with the SSR rendering result (/electronics-spa/en/USD/Brands/all/c/brands)",
    / ... /
  }
  ```
- refresh the page and check in logs whether a page has been taken from the cache
  ```
  {
    "message": "Render from cache (/electronics-spa/en/USD/Brands/all/c/brands)",
  / ... /
  }
  ```
- go to `http://localhost:4200/electronics-spa/en/USD/`. As the `cacheSize` has been set to `1`, the cached home page was overwritten by a page rendered for `/electronics-spa/en/USD/Brands/all/c/brands`. Verify in logs if the homepage was rendered:
  ```
  {
    "message": "Request is resolved with the SSR rendering result (/electronics-spa/en/USD/)",
    / ... /
  }
  ```

closes [CXSPA-7778](https://jira.tools.sap/browse/CXSPA-7778)
